### PR TITLE
[Python] Add support for Python=3.11

### LINKED ISSUE #1776: Python=3.11 is now out and about
so we should test and include it in our envs :snake:

### PR DESCRIPTION
**Pull request by valeriupredoi**
## Description

This is **THERE** :grin: 

Steps I took:

- ~`esmpy` - complicated, see https://github.com/conda-forge/esmpy-feedstock/pull/57~
- ~`iris` (via `netCDF4` repo patch) - this will take a while to be unpatched, depends how fast the `iris` folk will understand and fix the issues from `netCDF4 > 1.6.0`~
- ~`iris-esmf-regrid` (needs `esmpy`)~

As of 10 May, 2023: we have support for Python=3.11:

- ~`prospector` still not supporting Py=3.11 - this is work in progress by me, see https://github.com/ESMValGroup/ESMValCore/pull/1832#issuecomment-1516125341~
- this is still a valid issue, but not affecting us *too much*: `sphinx` needs to be unpinned so 5.0 gets pulled, 6+ has issues with the rtd theme, ~see example env file below; the issue is a pin on `docutils` the sphinx_rtd_theme has in their meta file, see https://github.com/conda-forge/sphinx_rtd_theme-feedstock/issues/28~ - it looks like it'll have to stay like this for a while:  `sphinx >=6` can not be installed because `sphinx_rtd_theme` are pinning `docutils<0.19` which is the only practical version that supports `python=3.11`; see https://github.com/readthedocs/sphinx_rtd_theme/issues/1323 and github.com/conda-forge/sphinx_rtd_theme-feedstock/issues/28
- ~RTD build fails related to this https://github.com/conda-forge/esmf-feedstock/issues/91 since the RTD mechanism does not activate an env, but rather, uses the `(base)` env (silly!)~ neatly sorted out by @zklaus with a bit of magic :magic_wand: but the RTD interface is still obtuse, at least docs build now
- ~the tests fail massively due to geopy's geoextractor known HTTP request timeout; this should be fixed by https://github.com/ESMValGroup/ESMValCore/pull/2002 - @zklaus found out that Nominatim (the locator engine) should not be used in CIs - a bit of a letdown (not Klaus' work, which is great, but the robustness of Nominatim), for now I am pytest-skipping `tests/integration/preprocessor/_regrid/test_extract_location.py`~ fixed by merging `main` that now contains @schlunma 's mockingbird from #2005 

Apart from that above, all goes spiffy!

Closes #1776 and #2001

### CODE REVIEW

**Comment by valeriupredoi** [comment]
an absolute fail so far, only Codacy passed :rofl: Give it another couple of months :grin:

**Comment by bouweandela** [comment]
I don't know about that, bit indeed it looks like it's over a year old.

**Comment by valeriupredoi** [comment]
> I don't know about that, but indeed it looks like it's over a year old.

yeah the ESMPY itself is one year old but xESM is actually not maintained (there was that guy that went under a bus), so am not sure how to ping these folks - am gonna try and open a PR in their feedstock update the build for 3.11

**Comment by valeriupredoi** [comment]
actually there is a PR awaiting for exactly that (albeit not in good shape) https://github.com/conda-forge/esmpy-feedstock/pull/57 - I asked if we can get done linux64-only PR to start supporting 3.11

**Comment by valeriupredoi** [comment]
OK I've managed to get more work done on this, with the new iris 3.4.1 now (indirectly supporting 3.11, via the unpin of netCDF4), the only two issues in the environment are:

- the pin on `esmpy <8.4` (I'll investigate the associated issue)
- and, as always, `prospector` - I maight just go open a PR on prospector's feedstock

**Comment by valeriupredoi** [comment]
right - so the issue with esmpy >=8.4 via iris-esmf-regrid is nicely raised here https://github.com/SciTools-incubator/iris-esmf-regrid/issues/227 but that issue it's been there for a wee while

**Comment by valeriupredoi** [comment]
PR-ed iris-esmf-regrid, next stop is drop by prospector, I am the "Python 3.11 Spanish Inquisition" :grin:

**Comment by valeriupredoi** [comment]
prospector don't work with Python 3.11 coz requirements-detector don't work - it stops at Python 3.10 - gonna go inquire there "excuse me, sir, can we Python 3.11 you please" :grin:

**Comment by dhohn** [comment]
> prospector don't work with Python 3.11 coz requirements-detector don't work - it stops at Python 3.10 - gonna go inquire there "excuse me, sir, can we Python 3.11 you please" 😁

I went down that rabbit hole a while back for mac ARM versions as well. The thing with the requirements-detector is that it has binary builds. A noarch version would also make py3.11 easier.

**Comment by valeriupredoi** [comment]
@dhohn right you are indeed! Let's see what the folks there say...I am tempted to just open a PR, next week

**Comment by valeriupredoi** [comment]
ah but there actually is a PR to update requirements-detector to Py311 - I should have checked yesterday https://github.com/conda-forge/requirements-detector-feedstock/pull/15

**Comment by zklaus** [comment]
Ok, I think I kind-of got it, see https://github.com/ESMValGroup/ESMValCore/pull/2003. There is another problem then, but maybe you've seen that before?

**Comment by valeriupredoi** [comment]
just look at all those geopy.geoextractor test fails! Hole cow :cow:

**Comment by valeriupredoi** [comment]
OK @bouweandela @zklaus - we are almost there wrt supporting Python=3.11 and boy the road here was not an easy one! I'd like you input on a couple things please, see the PR description https://github.com/ESMValGroup/ESMValCore/pull/1832#issue-1466762628

- prospector is a spanner in the works, it flat out refuses to support Python 3.11 - shall we make it an optional dependency? We even have an issue for that
- the test that does the location extraction is currently skipped (otherwise it'd fail a zillion times) - see https://github.com/ESMValGroup/ESMValCore/issues/1982 and attached PR that builds a custom marker for it; Klaus says they don't allow CI runs, so - are we just gonna skip that test always then?

I'd appreciate it if could have a review please - if we decide on those two matters above then I think this is ready to go. Of course, the plot will thicken when we'd have to see how well Tool plays with Python=3.11 :laughing:

File: environment.yml [hunk 1]
```diff
@@ -45,7 +45,12 @@ dependencies:
   - autodocsumm>=0.2.2
   - ipython
   - nbsphinx
-  - sphinx>=6.1.3
+  # sphinx >=6 can not be installed because
+  # sphinx_rtd_theme are pinning docutils<0.19 which is the only
```
Comment (bouweandela) [inline environment.yml:49]: Maybe we could use the theme suggested by @zklaus in another issue (can't find it right now)?

File: doc/conf.py [hunk 2]
```diff
@@ -33,6 +33,12 @@
 
 # This is used for linking and such so we link to the thing we're building
 rtd_version = os.environ.get("READTHEDOCS_VERSION", "latest")
+if on_rtd:
+    rtd_project = os.environ.get("READTHEDOCS_PROJECT")
+    rtd_conda_prefix = f"/home/docs/checkouts/readthedocs.org/user_builds/{rtd_project}/conda/{rtd_version}"
+    os.environ["ESMFMKFILE"] = f"{rtd_conda_prefix}/lib/esmf.mk"
+    os.environ["PROJ_DATA"] = f"{rtd_conda_prefix}/share/proj"
+    os.environ["PROJ_NETWORK"] = "OFF"
```
Comment (bouweandela) [inline doc/conf.py:41]: Some comments/references to issues explaining why these environmental variables are needed would be nice

File: tests/integration/preprocessor/_regrid/test_extract_location.py [hunk 3]
```diff
@@ -5,10 +5,12 @@
 import numpy as np
 from iris.coords import CellMethod, DimCoord
 
+import pytest
 import tests
 from esmvalcore.preprocessor import extract_location
 
 
+@pytest.mark.skip(reason="This results in multiple failed HTTP requests.")
```
Comment (bouweandela) [inline tests/integration/preprocessor/_regrid/test_extract_location.py:13]: This has now been solved in #2005

File: environment.yml (hunk 1 above)
Comment (zklaus) [reply environment.yml:49]: I suggested that as a remedy for this problem here in #2004, which is a PR targeting this branch.

Comment (valeriupredoi) [reply environment.yml:49]: gents, yes - but let's stick to the scope of this PR - #2004 (or, its previous incarnation) was merged here - through which good old Klaus fixed the RTD crappe that gave me plenty of headaches. Let's see about the new theme in a separate PR though, chances are that may be merged before this one anyway :grin:

File: setup.py [hunk 4]
```diff
@@ -95,7 +99,7 @@
         'docformatter',
         'isort',
         'pre-commit',
-        'prospector[with_pyroma,with_mypy]!=1.1.6.3,!=1.1.6.4',
+        # 'prospector[with_pyroma,with_mypy]!=1.1.6.3,!=1.1.6.4',
```
[the remark this answers is not in the document]
Comment (valeriupredoi) [reply setup.py:102]: I am totally on board with prospector! Let me try this first - I am gonna go see what's holding it up from supporting Py11, so go upstream, then, depending on what's happening there, we see about it here. Now, after @schlunma mocked the location extraction test, we should be pretty close and clean on this PR

File: doc/conf.py (hunk 2 above)
Comment (valeriupredoi) [reply doc/conf.py:41]: good point here, @zklaus you mind doing the service, bud?

File: tests/integration/preprocessor/_regrid/test_extract_location.py (hunk 3 above)
Comment (valeriupredoi) [reply tests/integration/preprocessor/_regrid/test_extract_location.py:13]: I am onto it now

Comment (valeriupredoi) [reply tests/integration/preprocessor/_regrid/test_extract_location.py]: done

File: environment.yml (hunk 1 above)
Comment (zklaus) [reply environment.yml:49]: We can do that, though, as you noted, sphinx_rtd_theme breaks Python 3.11 compatibility.

Comment (valeriupredoi) [reply environment.yml:49]: well, not for sphinx<6 - can we live with sphinx=5 for a while til we get our new theme up and spick and span?

Comment (zklaus) [reply environment.yml:49]: Hm. That breaks (at least) the API docs for `esmvalcore.cmor.check.CheckLevels`, though maybe we can live without them for a bit?

Comment (valeriupredoi) [reply environment.yml:49]: yeah was thinking that too

**Comment by valeriupredoi** [comment]
@bouweandela @zklaus heads up I've started working on this, to bump `prospector` to a modern version on conda, and hopefully get support for Py311 working there, feel free to visit https://github.com/conda-forge/prospector-feedstock/pull/39

**Comment by valeriupredoi** [comment]
OK folks here's how things stand now (April 20, 2023) wrt to `prospector`:

- I opened a PR to bump `prospector` to a modern version on conda, and hopefully get support for Py311 working there, https://github.com/conda-forge/prospector-feedstock/pull/39
- that PR is currently failing them tests since `requirements-detector` can not be installed from conda with a modern enough version and satisfying their pin for it; apart from that all else seems to be fine
- as such, I opened a PR to update the `requirements-detector` conda package at https://github.com/conda-forge/requirements-detector-feedstock/pull/18
- but that is a bit sick too since `requirements-detector` depends on `poetry-semver` which is now archived, so what I'll do I'll propose a PR that replaces it with `semver`, see https://github.com/landscapeio/requirements-detector/issues/43

**Comment by bouweandela** [comment]
That's turning into a nice bit of yak shaving you're doing there V! Much appreciated!

**Comment by valeriupredoi** [comment]
"yak shaving" :rofl: A fairly furry yak too, bud :grin:

**Comment by bouweandela** [comment]
https://www.youtube.com/watch?v=AbSehcT19u0

**Comment by valeriupredoi** [comment]
OK - https://github.com/conda-forge/prospector-feedstock/pull/39 has now been merged and that means we should be fine for Py311 here - let's see how it goes that WD40 :grin:

**Comment by valeriupredoi** [comment]
@bouweandela @zklaus looky-look! We have us full support for Python 3.11, gents, no corners cut, no prospectors left behind :grin: I'll now clean up this and we are ready to plug it in :beers:

File: doc/conf.py (hunk 2 above)
Comment (bouweandela) [reply doc/conf.py:46]: Thanks!

**Review by bouweandela — APPROVED** [review]
Looks good to me!

File: environment.yml (hunk 1 above)
Comment (bouweandela) [reply environment.yml]: Done

**Pull request merged**